docs/design/features/source-generator-pinvokes.md

```diff
@@ -0,0 +1,232 @@
+
+The user experience of the built-in generation initially appears ideal, but there are several negative consequences that make the system costly in the long term:
+
+* Bug fixes in the marshaling system require an update to the entire runtime.
+* New types require enhancements to the marshaling system for efficient marshal behavior.
+  * [`ICustomMarshaler`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.icustommarshaler) incurs a substantial performance penalty.
+* Once a marshaling bug becomes expected behavior the bug is difficult to fix. This is due to user reliance on shipped behavior and since the marshaling system is built into the runtime there aren't ways to select previous or new behavior.
+  * Example involving COM marshaling: https://github.com/dotnet/coreclr/pull/23974.
+* Debugging the auto-generated marshaling IL Stub is difficult for runtime developers and close to impossible for consumers of P/Invokes.
+
+This is not to say the P/Invoke system should be completely redesigned. The current system is heavily used and its simplicity for consuming native assets is a benefit. Rather this new mechanism is designed to provide a way for marshaling code to be generated by an external tool but integrate with `DllImportAttribute` in a way that isn't onerous on current .NET developers.
+
+The [Roslyn Compiler](https://github.com/dotnet/roslyn) team is working on a [Source Generator feature][source_gen_link] that will allow the generation of additional source files that can be added to an assembly during the compilation process - the runtime generation IL Stubs is an in-memory version of this scenario.
+
+**Note** This proposal is targeted at addressing P/Invoke improvements but could be adapted to work with COM interop utilizing the new [`ComWrappers`][comwrappers_link] API.
+
+### Requirements
+
+* [Source generators][source_gen_link]
+  * Branch: https://github.com/dotnet/roslyn/tree/features/source-generators
+
+## Design
+
+Using Source Generators is focused on integrating with existing uses of `DllImportAttribute` from an invocation point of view (i.e. callsites should not need to be updated). The idea behind Source Generators is that code for some scenarios can be precomputed using user declared types and logic thus avoiding the need to generator code at runtime. The desire is then to provide a way that existing code can continue to function but can be modified in a way that allows it to leverage this new compiler feature.
```

Comment:
> The desire is then to provide a way that existing code can continue to function but can be modified in a way that allows it to leverage this new compiler feature.

How do you envision exiting code continues working here while taking advantage of the newly generated code. The existing code is PInvoke which is an `extern` method. Is the idea that the runtime will change to forward the PInvoke stub to the generated helpers?

Reply:
Well, ideally in my mind, for this situation, the user written function remains a valid function (for the purposes of reflection, etc), but the body of the function simply calls some other method that provides the implementation. I don't know if this is practical at all, but … you did ask. :)

Reply:
I believe this would be addressed with https://github.com/dotnet/csharplang/issues/3301